app/src/org/commcare/sync/FirebaseMessagingDataSyncer.java

```diff
@@ -0,0 +1,334 @@
+import static org.commcare.utils.FirebaseMessagingUtil.removeServerUrlFromUserDomain;
+
+import android.content.Context;
+import android.widget.Toast;
+
+import org.commcare.CommCareApplication;
+import org.commcare.activities.CommCareActivity;
+import org.commcare.activities.EntityDetailActivity;
+import org.commcare.activities.EntitySelectActivity;
+import org.commcare.activities.FormEntryActivity;
+import org.commcare.activities.MenuActivity;
+import org.commcare.activities.StandardHomeActivity;
+import org.commcare.models.AndroidSessionWrapper;
+import org.commcare.preferences.HiddenPreferences;
+import org.commcare.preferences.ServerUrls;
+import org.commcare.services.CommCareSessionService;
+import org.commcare.services.FCMMessageData;
+import org.commcare.session.CommCareSession;
+import org.commcare.tasks.DataPullTask;
+import org.commcare.tasks.ResultAndError;
+import org.commcare.tasks.templates.CommCareTask;
+import org.commcare.tasks.templates.CommCareTaskConnector;
+import org.commcare.util.LogTypes;
+import org.commcare.utils.SyncDetailCalculations;
+import org.commcare.views.dialogs.PinnedNotificationWithProgress;
+import org.javarosa.core.model.User;
+import org.javarosa.core.services.Logger;
+import org.javarosa.core.services.locale.Localization;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FirebaseMessagingDataSyncer implements CommCareTaskConnector {
+
+    private Context context;
+
+    public FirebaseMessagingDataSyncer(Context context) {
+        this.context = context;
+    }
+
+    public enum AppNavigationStates {
+        FORM_ENTRY,
+        ENTITY_SELECTION,
+        ENTITY_DETAIL,
+        MENU,
+        HOME_SCREEN,
+        OTHER
+    }
+    private CommCareTask currentTask = null;
+    private PinnedNotificationWithProgress<DataPullTask.PullTaskResult> mPinnedNotificationProgress = null;
+
+    /**
+     * If we land here is because there is a data payload and the action there is to sync
+     * 1) Check if there is an active session.
+     * - If yes, this will lead to further steps to attempt to trigger a sync
+     * - If not, a sync will be scheduled after the next successful login
+     * 2) Ensure that the sync is only triggered for the intended 'recipient' of the message
+     */
+    public void syncData(FCMMessageData fcmMessageData) {
+        // Abort if FCM Message data is null
+        if (fcmMessageData == null){
+            disablePendingSyncs();
+            return;
+        }
+
+        if (!CommCareApplication.isSessionActive()) {
+            //  There is no active session at the moment, proceed accordingly
+            // TODO: Decide whether to only trigger the Sync when the 'recipient' of the message logs in
+            //  or anyone, in case multiple users are sharing the same device
+            // TODO: Decide whether to check if when there is no active session, the recipient has ever
```

Comment:
In case a sync request comes through for a user that hasn't yet logged in in the device. Considering that during the first login process a sync is triggered, I don't think we need to trigger any pending sync. A situation when this can be applicable is when a user clears user data and logs in again, the token won't change and during the now "first login" we might have pending sync requests.

Reply:
Yes, I like that